PhysicsTools/HepMCCandAlgos/plugins/GenPUProtonProducer.cc

```diff
@@ -0,0 +1,268 @@
+  abortOnUnknownPDGCode_( cfg.getUntrackedParameter<bool>( "abortOnUnknownPDGCode", true ) ),
+  bunchList_( cfg.getParameter<vector<int> >( "bunchCrossingList" ) ),
+  minPz_( cfg.getParameter<double>( "minPz" ) ) 
+{
+  produces<GenParticleCollection>();
+  mixToken_ = consumes<CrossingFrame<HepMCProduct> >( InputTag(cfg.getParameter<std::string>( "mix" ),"generatorSmeared") );
+}
+
+GenPUProtonProducer::~GenPUProtonProducer() { }
+
+void GenPUProtonProducer::produce( Event& evt, const EventSetup& es ) {
+
+   if ( !convertParticle_.initialized() ) {
+     ESHandle<HepPDT::ParticleDataTable> pdt_handle;
+     es.getData( pdt_handle );
+     HepPDT::ParticleDataTable const& pdt = *pdt_handle.product();
+     convertParticle_.init( pdt ); 
+   }
+
+   unsigned int totalSize = 0;
+   unsigned int npiles = 1;
+
+   Handle<CrossingFrame<HepMCProduct> > cf;
+   evt.getByToken(mixToken_,cf);
+   std::auto_ptr<MixCollection<HepMCProduct> > cfhepmcprod( new MixCollection<HepMCProduct>( cf.product() ) );
+   npiles = cfhepmcprod->size();
+
+   LogDebug("GenPUProtonProducer") << " Number of pile-up events : " << npiles << endl;
+
+   for(unsigned int icf = 0; icf < npiles; ++icf){
+      LogDebug("GenPUProtonProducer") << "CF " << icf << " size : " << cfhepmcprod->getObject(icf).GetEvent()->particles_size() << endl;
+      totalSize += cfhepmcprod->getObject(icf).GetEvent()->particles_size();
+   }
+   LogDebug("GenPUProtonProducer") << "Total size : " << totalSize << endl;
+
+   // Loop over pile-up events
+   MixCollection<HepMCProduct>::MixItr mixHepMC_itr;
+   unsigned int total_number_of_protons = 0;
+   size_t idx_mix = 0;
+   for( mixHepMC_itr = cfhepmcprod->begin() ; mixHepMC_itr != cfhepmcprod->end() ; ++mixHepMC_itr, ++idx_mix ){
+      int bunch = mixHepMC_itr.bunch();
+      if( find( bunchList_.begin(), bunchList_.end(), bunch ) != bunchList_.end() ){
+	 auto event = (*mixHepMC_itr).GetEvent();
+	 // Find protons
+	 unsigned int number_of_protons = 0;
+	 for( auto p = event->particles_begin() ; p != event->particles_end() ; ++p ){
+	    if( select_(*p, minPz_) ) ++number_of_protons;
+	 }
+	 LogDebug("GenPUProtonProducer") << "Idx : " << idx_mix << " Bunch : " << bunch << " Number of protons : " << number_of_protons << endl;
+	 total_number_of_protons += number_of_protons;
+      }
+   }
+   LogDebug("GenPUProtonProducer") << "Total number of protons : " << total_number_of_protons << endl;
+
+   // Initialise containers
+   const size_t size = total_number_of_protons;
+   vector<const HepMC::GenParticle *> particles( size );
+
+   auto candsPtr = std::make_unique<GenParticleCollection>(size);
+   auto barCodeVector = std::make_unique<vector<int>>(size);
+   ref_ = evt.getRefBeforePut<GenParticleCollection>();
+   GenParticleCollection& cands = *candsPtr;
+
+   size_t idx_particle = 0;
+   idx_mix = 0;
+   // Fill collection
+   for( mixHepMC_itr = cfhepmcprod->begin() ; mixHepMC_itr != cfhepmcprod->end() ; ++mixHepMC_itr, ++idx_mix ){
```

Comment:
Why do you need two loops over the genParticles? It looks like you could just push back the protons onto the particle vector without needing to know how big it will eventually be.